File: tests/native/public-ip-classifier/PublicAddressPolicyProof.java
package org.nostrocket.hypergolic.host;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class PublicAddressPolicyProof {
  public static void main(String[] args) throws Exception {
    int checks = 0;
    for (String line : Files.readAllLines(Path.of(args[0]))) {
      if (line.isBlank() || line.startsWith("#")) continue;
      String[] fields = line.split("\\t", -1);
      boolean expected = Boolean.parseBoolean(fields[2]);
      boolean actual;
      if (fields[0].equals("LITERAL")) actual = PublicAddressPolicy.isPublicLiteral(fields[1]);
      else if (fields[0].equals("ANSWERS")) {
        List<String> answers = fields[1].isEmpty() ? List.of() : new ArrayList<>(Arrays.asList(fields[1].split(",", -1)));
        actual = PublicAddressPolicy.allPublicAnswers(answers);
      } else if (fields[0].equals("BYTES")) {
        byte[] address = new byte[fields[1].length() / 2];
        if (fields[1].length() % 2 != 0) throw new AssertionError("Odd-length byte vector: " + line);
        for (int i = 0; i < address.length; i++) address[i] = (byte) Integer.parseInt(fields[1].substring(i * 2, i * 2 + 2), 16);
        actual = PublicAddressPolicy.accepts(address);
      } else throw new AssertionError("Unknown vector type: " + fields[0]);
      if (actual != expected) throw new AssertionError("Vector " + (checks + 1) + " failed: " + line + " actual=" + actual);
      checks++;
    }
    System.out.println("Android/JVM classifier vectors passed: " + checks);
  }
}

File: tests/native/public-ip-classifier/PublicAddressPolicyProof.swift
import Foundation

@main
struct PublicAddressPolicyProof {
  static func main() throws {
    let path = CommandLine.arguments[1]
    let lines = try String(contentsOfFile: path, encoding: .utf8).components(separatedBy: .newlines)
    var checks = 0
    for line in lines where !line.isEmpty && !line.hasPrefix("#") {
      let fields = line.components(separatedBy: "\t")
      guard fields.count == 3 else { throw ProofFailure.invalidVector(line) }
      let expected = fields[2] == "true"
      let actual: Bool
      switch fields[0] {
      case "LITERAL": actual = PublicAddressPolicy.isPublicLiteral(fields[1])
      case "ANSWERS":
        let answers = fields[1].isEmpty ? [] : fields[1].components(separatedBy: ",")
        actual = PublicAddressPolicy.allPublicAnswers(answers)
      case "BYTES":
        guard fields[1].count.isMultiple(of: 2) else { throw ProofFailure.invalidVector(line) }
        var bytes = [UInt8]()
        var cursor = fields[1].startIndex
        while cursor < fields[1].endIndex {
          let end = fields[1].index(cursor, offsetBy: 2)
          guard let value = UInt8(fields[1][cursor..<end], radix: 16) else { throw ProofFailure.invalidVector(line) }
          bytes.append(value)
          cursor = end
        }
        actual = PublicAddressPolicy.accepts(Data(bytes))
      default: throw ProofFailure.invalidVector(line)
      }
      guard actual == expected else { throw ProofFailure.failed(checks + 1, line, actual) }
      checks += 1
    }
    print("iOS/Swift classifier vectors passed: \(checks)")
  }
}

enum ProofFailure: Error, CustomStringConvertible {
  case invalidVector(String)
  case failed(Int, String, Bool)
  var description: String {
    switch self {
    case .invalidVector(let line): return "Invalid vector: \(line)"
    case .failed(let number, let line, let actual): return "Vector \(number) failed: \(line) actual=\(actual)"
    }
  }
}
